jsonalchemy/wrappers.py

```diff
@@ -34,98 +35,150 @@ def validate(data, schema, *args, **kwargs):
     }
     return Draft4Validator(schema=schema, types=types).validate(data)
 
 
 def wrap(value, value_schema):
+
+    if isinstance(value, JSONBase):
+        return value
+
+    types = {
+        dict: ('object', JSONDict),
+        list: ('array', JSONList),
+        str: ('string', JSONString),
+        unicode: ('string', JSONString),
+        # TO DO integer vs number
+        int: ('integer', JSONNumber),
+        float: ('number', JSONNumber),
+        bool: ('boolean', JSONBool)
+    }
+
+    type = types.get(value.__class__, (None,))
+
     if value_schema:
         value_type = value_schema.get('type', '')
-        if value_type == 'object':
-            if not isinstance(value, JSONDict):
-                return JSONDict(value, value_schema)
-        elif value_type == 'array':
-            if not isinstance(value, JSONList):
-                return JSONList(value, value_schema)
-    return value
+        if not isinstance(value_type, list):
+            if type[0] == value_type:
+                return type[1](value, value_schema)
+        else:
+            if type in value_type:
+                # For example 'number' in ['number', 'list']
+                return type[1](value, value_schema)
+            else:
+                raise TypeError
+    elif value and not isinstance(value, JSONBase):
+        return type[1](value, value_schema)
 
 
-class JSONDict(collections.Mapping):
+class JSONBase(object):
 
-    def __init__(self, mapping=None, schema=None):
+    def __init__(self, schema=None):
         self.schema = schema or {}
-        mapping = mapping or {}
 
-        properties = self.schema.get('properties', {})
-        self.dict = {
-            name: wrap(value, properties.get(name, None))
-            for name, value in iteritems(mapping)
-        }
+    @property
+    def validation(parent):
 
-        validate(self.dict, self.schema)
+        class JSONValidation(object):
+            def __init__(self, json):
+                self._json = json
+                self._copy = json.__class__(json, json.schema)
 
-    def __iter__(self):
-        return iter(self.dict)
+            def __enter__(self):
+                return self._copy
 
-    def __len__(self):
-        return len(self.dict)
+            def __exit__(self, type, value, traceback):
+                if not type:
+                    # There was no exception
+                    validate(self._copy,
+                             self._json.schema)
+                    self._json.update(self._copy)
 
-    def __contains__(self, value):
-        return value in self.dict
+        return JSONValidation(parent)
 
-    def __getitem__(self, name):
-        return self.dict[name]
+    def validate(self):
+        return validate(self, self.schema)
 
-    def __setitem__(self, name, value):
-        has_key = name in self.dict
-        old_value = self.dict.get(name, None)
 
+class JSONDict(dict, JSONBase):
+
+    def __new__(self, mapping=None, schema=None):
```

Comment:
`__new__` takes `cls` (class) as first argument and not `self` (https://docs.python.org/2/reference/datamodel.html#object.__new__).